Added user autocomplete from ldap in create view and a parameter to enable it

AndreScara11 opened the pull request:
Implemented the possibility to autocomplete email and username from ldap when create a new user

maxxer commented on src/User/resources/views/admin/_user.php:
```diff
@@ -13,8 +13,60 @@
  */
+
+use Da\User\Dictionary\UserSourceType;
+use kartik\typeahead\Typeahead;
```
This is not a requirement of the module, projects without this module installed will receive an error. Must be added in `composer.json`. However, considering the restricted usage of this change, I'd rather avoid adding as a global requirement. Maybe we can avoid adding it to `composer.json`, we could make a library check during usuario's bootstrap. If the search flag is enabled, check if this class is defined.

maxxer commented on src/User/Module.php:
```diff
@@ -32,6 +32,10 @@ class Module extends BaseModule
      */
     public $numberSessionHistory = false;
+    /**
+     * @var bool Enable LDAP to sync local users with LDAP users
```
Isn't it for search?

Please add documentation for this additional parameter in the `docs` folder as well. Clearly explain the purpose and the requirements for this to work.

maxxer commented on src/User/Dictionary/UserSourceType.php:
```diff
@@ -0,0 +1,37 @@
+    {
+        return ArrayHelper::getValue(static::all(), $key);
+    }
+
```
Please run cs-fixer

maxxer commented on src/User/resources/views/admin/_user.php:
```diff
@@ -13,8 +13,60 @@
+use yii\helpers\Url;
+
 ?>
+<?php $source = Yii::$app->request->get('source') ?: $user->source ?>
```
Almost every statement has open/closing php tags. Let's make it pure php

maxxer commented on src/User/resources/views/admin/_user.php:
```diff
@@ -13,8 +13,60 @@
+    <?php if (Yii::$app->getModule('user')->searchUsersInLdap && $source==UserSourceType::LDAP): ?>
+        <?= $form->field($user, 'source')->dropDownList(UserSourceType::all(), ['value' => $source]); ?>
+        <?= $form->field($user, 'email')->widget(Typeahead::class, [
+                'options' => ['placeholder' => Yii::t('mis', 'Filter as you type ...')],
```
Wrong language category.

maxxer replied:
Please run message extraction before the next commit.

maxxer commented on src/User/resources/views/admin/_user.php:
```diff
@@ -13,8 +13,60 @@
+        ?>
+        <?= $form->field($user, 'username')->hiddenInput()->label(false) ?>
+    <?php else: ?>
+        <?= $form->field($user, 'source')->dropDownList(UserSourceType::all(), ['value' => $source]); ?>
```
How does this parameter affect the user creation? I see no reference to `$source` anywhere else.

maxxer reviewed (changes requested):
Thank you for your contribution. Please review the suggested changes, don't forget to extract messages and run `php-cs-fixer`.

maxxer commented:
@schmunk42 @eluhr since you use LDAP extension, do you mind having a look as well? thank you

maxxer commented on docs/install/configuration-options.md:
```diff
@@ -14,6 +14,32 @@ Number of expired storing records `session history`, values:
+
+#### ldapUserAttributes (Type: `array`)
+
+This array maps the user attributes to sync from ldap
```
Please use LDAP with consistency. Few lines above is full capital

maxxer commented on docs/install/configuration-options.md:
```diff
@@ -14,6 +14,32 @@ Number of expired storing records `session history`, values:
+
+#### ldapProfileAttributes (Type: `array`)
+
+This array maps the profile attributes to sync from ldap
```
Same as above

maxxer commented on src/User/Bootstrap.php:
```diff
@@ -154,6 +154,12 @@ function () use ($model) {
 
+            if (Yii::$app->getModule('user')->searchUsersInLdap) {
+                if (!class_exists('kartik\typeahead\Typeahead')) {
+                    throw new InvalidConfigException('The kartik-v/yii2-widget-typeahead library must be defined when searchUsersInLdap is true.');
```
_must be **installed**_ rather than _defined_

TonisOrmisson replied:
also here https://github.com/2amigos/yii2-usuario/pull/543/files#diff-3af9b9b60ce0ece19bdfd796fed5974821179bdc8da2278e7f5d0c8b4b6671eeR21 is a new dependency not installed, nor documented

AndreScara11 commented:
Last changes:

- Used Selectizetextinput instead of Typeahead
- Added property `ldapUid` to User model to handle the ldap users without the username
- Changed ldap params names
- Updated docs

maxxer commented on docs/install/configuration-options.md:
````diff
@@ -18,13 +18,13 @@ Number of expired storing records `session history`, values:
 ```
 
-#### ldapUserAttributes (Type: `array`)
+#### LDAPUserAttributes (Type: `array`)
````
I meant capitalization in the description, not in the attribute name! We have no config param starting with a capital. Please restore the previous attribute name, and fix the description at line 29

AndreScara11 commented:
Done, sorry for the misunderstanding.

AndreScara11 commented:
Have you had a chance to review?

eluhr commented:
> @schmunk42 @eluhr since you use LDAP extension, do you mind having a look as well? thank you

@maxxer  since you asked for my opinion on this, here is my assessment. I find it somewhat critical to wire the LDAP so closely to the package. I think the whole thing would be better off in a separate package, which overwrites the required views and controller actions or works with the existing events at this point.
